terraso_backend/apps/auth/services.py

```diff
@@ -250,3 +250,60 @@ def track_signup(self, auth_provider: str, req) -> None:
                 return
         props = {"service": auth_provider}
         self.track_event(event_name, user_agent, ip_address, self.EVENT_URL, props)
+
+
+class TokenExchangeException(Exception):
+    def __init__(self, message: str, error_type: str):
+        super().__init__(message)
+        self.message = message
+        self.error_type = error_type
+
+
+class TokenExchangeService:
+    def __init__(self, token, jwks_uri, client_id, provider_name):
+        self.token = token
+        self.jwks_uri = jwks_uri
+        self.client_id = client_id
+        self.provider_name = provider_name
+
+    @classmethod
+    def from_payload(cls, payload, settings):
+        provider_name = payload["provider"]
+        token = payload["jwt"]
+        provider = settings.JWT_EXCHANGE_PROVIDERS[provider_name]
+        if "url" not in provider or "client_id" not in provider:
+            raise TokenExchangeException(
+                f"provider {provider_name} is missing config variables", "bad_config"
+            )
+        return cls(
+            token=token,
+            jwks_uri=provider["url"],
+            client_id=provider["client_id"],
+            provider_name=provider_name,
+        )
+
+    @staticmethod
+    def _get_signing_key(token, provider_url):
+        # fetch jwks
+        jwks_client = jwt.PyJWKClient(provider_url)
+        return jwks_client.get_signing_key_from_jwt(token)
+
+    @staticmethod
+    def _verify_payload(token, signing_key, client_id):
+        algorithms = [signing_key._jwk_data.get("alg", "RS256")]
```

Comment:
Question: Why is RS256 the default algorithm? it seems that we are using a different default for our current auth process, see: https://github.com/techmatters/terraso-backend/blob/main/terraso_backend/config/settings.py#L283

Reply:
I set this as the default because that is what was using by Google. Looking at this again, it probably makes more sense to throw an exception if the key is not present - it will be easier to read the logs and figure out what is going on.